### ISSUE TITLE
Replace the game's procedural sound effects with locally generated AI audio -- fanfares, fireworks, ball impacts

### ISSUE THREAD

**Issue by AntoninPrazsky** [AntoninPrazsky/BS3D#482]
Owner note (2026-09-19), translated from Czech: **"The game's sounds really need to be replaced with generative AI — victory fanfares, fireworks, ball sounds. All of it would be more believable generated. Need to find the best local AI model that can do this on the 6900 XT, CPU offload is fine."**

## What is there today

Every sound effect in the game is procedural — synthesised from raw PCM at load in `Game/Audio/ProceduralAudio.cs` (see "The sound" in `docs/game-feedback.md`). There is no audio file anywhere in the sound-effect path. This mirrors exactly where the game's *music* stood before #443.

## This is the music changeover's own trajectory, one step behind it

- **#187** (2026-09-14) originally ruled out *any* AI-generation service for the game, for music, partly over the unsettled copyright question around pure AI output.
- **#443** (2026-09-16) reversed that ruling for music specifically: locally generated ACE-Step tracks now ship as `Game/Music/*.ogg` (#444 put them in Ogg Vorbis), replacing the procedural score outright, with the six procedural pieces kept only as an About-page easter egg.
- **#442** (open) is the sound-effects analogue of that research step, but scoped deliberately narrow — Stable Audio 3.0 Small-SFX as a **reference to listen to and measure** while hand-tuning the procedural bakes, with an explicit line that nothing generated ships and the owner would need to confirm this doesn't reopen #187's question before anything is installed.

**This note is the owner asking for the sound-effects equivalent of what #443 already did for music: generate and actually ship the audio**, not just use it as a design reference. That already implicitly answers #442's open question (does replacing #187's "no generation" stance for SFX conflict with it — no, #443 already reversed it for the same reasoning applied to music, and this note asks for the same reversal here).

## What to do

1. Find the best local model that can synthesize convincing game SFX on this machine — the RX 6900 XT has no ROCm support on Windows, so CPU-only (Ryzen 9 5900X) or CPU-offloaded inference is the expected path; #442's own research already named Stable Audio 3.0 Small-SFX (459M params, open weights, Windows/CPU-plausible) as one candidate worth testing for real generation quality, not just as a reference.
2. Generate the specific sounds this note calls out — victory fanfare, fireworks/burst, ball impact/attach sounds — and put them beside the existing procedural bakes for an ear comparison, the same way #443's masters were judged against the procedural score before the switch.
3. Follow #443's own measured constraints where they transfer (fixed sample rate, any generation-length ceiling, whatever the tool's own prompt-adherence quirks turn out to be) and the game's standing sound taste when judging candidates: warm and bass-heavy over strictly realistic, no harsh drive, no hiss, no high-frequency debris — pleasant over credible.
4. If a model is adopted, its generation tooling belongs in `Tools/` beside `MusicBake`, per #446's own note that the music pipeline still lives outside the repo at `C:\Users\panrd\AI`.

## Where to look

`Game/Audio/ProceduralAudio.cs`, "The sound" in `docs/game-feedback.md`, #442 (the reference-only research, now supersedable), #443/#444 (the music precedent this follows), #187 (the original ruling this reverses for sound the way #443 reversed it for music), and `.claude/skills/local-ai` for how the desktop's local models are already driven from this repo.

**Comment by AntoninPrazsky** [comment]
**Status 2026-09-21: the CPU pipeline stands, the real model is behind the owner's Hugging Face login, and the free base model renders noise.**

**What is gated and what is not** (HF API, checked): `stabilityai/stable-audio-3-small-sfx` — the adversarially post-trained SFX model, the one the README says to generate with — is `gated: auto`: an HF account, "I Accept" on the model page, and a token on this machine (there is none). `stabilityai/stable-audio-3-small-sfx-base` (the pre-trained checkpoint, "intended for fine-tuning") is not gated and is what this session could run.

**What is installed, outside the repository:** `C:\Users\panrd\AI\sfx\venv` (uv, Python 3.13, torch 2.7.1 **CPU**, transformers 5.17, the `stable-audio-3` library from GitHub — `stable-audio-tools` on PyPI wants Python <3.11), the base model in the HF cache (3.3 GB without `svd_bases.pt`), and `generate-sfx.py` + `prompts-482.json` (five prompts written to the game's taste: the cannon shot, a ball attaching, a group pop, a firework burst, a victory fanfare; a `.txt` sidecar per render with everything that shaped it).

**One trap on the way:** the base model's `model_config.json` points its T5Gemma text encoder at the *gated* repo (`repo_id: stabilityai/stable-audio-3-small-sfx`, `subfolder: t5gemma-b-b-ul2`) although the base repo ships the same folder. The script rewrites the conditioner's `repo_id` to the model's own repo, and the load is clean: 685 checkpoint keys, 0 missing, 0 unexpected, 567 M parameters with the autoencoder.

**Measured on the CPU (Ryzen 9 5900X, 12 threads), 50 steps, cfg 7, seed 11:** model load 9.4 s; **40–43 s a render** for 1.5–4 s of 44.1 kHz stereo (the time is per-step, not per-second-of-audio). **And every render is full-scale broadband noise:** 39–90 % of the samples at the clip level, RMS −0.3 to −3.3 dBFS with a flat envelope across the whole clip, spectral centroid 8.6–9.8 kHz, zero-crossing rate 0.31–0.41 — no attack, no decay, nothing that is a sound. Either the pre-trained base genuinely does not generate without the post-training (which is what its README implies), or the CPU attention fallback (`flash_attn not installed, disabling varlen attention`) breaks the conditioning; the loader check rules out a weight mismatch. Not worth more CPU hours: the comparison this issue asks for needs the real model.

**What the owner does next, ~2 minutes:** log in at huggingface.co, open `stabilityai/stable-audio-3-small-sfx` and accept the licence, then on the desktop:

```
C:\Users\panrd\AI\sfx\venv\Scripts\hf.exe auth login
cd C:\Users\panrd\AI\sfx && venv\Scripts\python.exe generate-sfx.py prompts-482.json --model small-sfx --out out\482-sfx --steps 8 --cfg 1
```

(8 steps at cfg 1 are the post-trained model's own defaults in the library; expect ~7 s a render on the CPU if the per-step cost holds.) An agent picks it up from there: the five renders beside the procedural bakes for the ear comparison, then the `Tools/` pipeline if a model is adopted.

**Licence, for the shipping decision:** Stability AI Community License — free below the revenue threshold, but commercial use requires registering at stability.ai/community-license, and the attribution clause covers "a product or service that uses any portion" of the Materials; whether generated audio counts is the owner's reading before anything generated ships. T5Gemma is redistributed under the Gemma terms.

**Comment by AntoninPrazsky** [comment]
**The real model runs, on the CPU, and the five renders are ready for the owner's ear** — `C:\Users\panrd\AI\sfx\out\482-sfx\*.wav` (44.1 kHz stereo float, a `.txt` beside each with the prompt, negative prompt, seed, steps, cfg and the numbers).

The owner logged the environment in (`hf auth login`, token in the user profile, nowhere in the repository) and accepted the licence; `stabilityai/stable-audio-3-small-sfx` downloaded (3.3 GB) and loaded clean. **8 steps, cfg 1 (the post-trained model's own defaults), seed 11, CPU only: 9.0 s to load, 5.6–6.3 s a render** — against 40 s for the base model's 50 steps of noise.

| Render | Replaces | What the numbers say |
|---|---|---|
| `cannon-shot-11` | `BakeShoot` | **A deep boom**: centroid 317 Hz, 99.9 % of the energy under 250 Hz, attack then a ~1 s decay to silence, no clipping. The taste the game asks for, on paper. |
| `ball-attach-11` | `BakeLanded` | Wrong on paper: three bright hits (centroid 9 kHz) and 5.3 % of the samples clipped — the prompt said a muted rubbery thud. Re-prompt or another seed. |
| `group-pop-11` | `BakeRelease` | Two clean pops with silence between, centroid 2.6 kHz, no clipping. |
| `firework-burst-11` | `BakeFireworkBurst` | A burst then a long falling tail, centroid 1.3 kHz, 30 % bass, no clipping. |
| `victory-fanfare-11` | the CLEARED fanfare | ~2.5 s of tonal content then a decay, centroid 2.2 kHz, no clipping. Whether it is a fanfare is for the ear. |

**Next, once the owner has listened:** for the sounds he likes, more seeds and prompt variants (`prompts-482.json`, `--count`-style sweeps are one line in the script), then loudness-matching against the procedural bakes the way `MusicBake --tracks` matches the music; the ones he rejects get a re-prompt. If a set is adopted, the generator moves into `Tools/` beside `MusicBake` as this issue asks, and the licence's registration and attribution clauses are read before anything ships.

**Comment by AntoninPrazsky** [comment]
**Owner's ear on round one (2026-09-21):** the base model's renders all distorted (they are noise — confirmed above); with the real model `ball-attach` distorted and the other four *readable but dull, weak and boring*. Two causes on my side: prompts written around "soft, gentle, clean" gave timid sounds, and nothing normalised the output (raw peaks 0.2–0.5).

**Round two** — `prompts-482-v2.json`, three punchier variants per sound ("punchy, satisfying, arcade, big, juicy"), three seeds each, `--count 3 --normalize -1` (peaks to −1 dBFS, raw peak and RMS kept in the sidecar), ~6 s a render on the CPU: 45 renders, **37 clean, and 8 of the 9 ball-attach renders clipped** (3–12 % of samples) — the model overdrives on short impact prompts ("one solid thock", "smack").

**Round three**, ball-attach only — "moderate volume", "gentle attack", negative "loud, distorted, clipping, aggressive": **9 of 9 clean** (raw peaks 0.14–0.82).

The lesson for the prompts, to go into the skill once a set is adopted: **for impacts tell the model the volume in words or it saturates; for everything else do not hold back on energy.** `generate-sfx.py` now takes `--count` and `--normalize` and writes an `index.html` of players beside the files, so a batch is listened to in a browser: `out\482-v2\index.html` and `out\482-attach\index.html`. Awaiting the owner's picks.

**Comment by AntoninPrazsky** [comment]
**Owner's picks, rounds two and three (2026-09-21):** cannon-shot a→seed 11, c→11 (b "all the same"); ball-attach a/b/c "all badly distorted, unusable" (matches the clipping), d→11, e→11, f→12; group-pop a→11, c→11 (b "all distorted" — the balloon prompts, dropped); firework-burst a→12, b→11, c→12; victory-fanfare a→11, b→12, c→12.

The picks are assembled per game sound in `C:\Users\panrd\AI\sfx\out\482-picks\index.html` (13 renders, sidecars beside them) for the one-per-sound choice. Note for the prompts: the ear hears distortion in `group-pop-b` although 0 % of its samples clip — bursting balloons are a texture the model renders harsh, so "balloon" is out of the vocabulary here.

**Comment by AntoninPrazsky** [comment]
**On `main` — merge e5ee560: the four chosen renders play in the game.**

- `Game/Sfx/shoot.ogg`, `landed.ogg`, `release.ogg`, `firework-burst.ogg` (the owner's picks `cannon-shot-c-11`, `ball-attach-e-11`, `group-pop-a-11`, `firework-burst-a-12`), written by the new `Tools/MusicBake --sfx <wav> <name>`: mono mix, peak 0.9, the tracks' encoder under a serial hashed from the name (byte-identical on a rerun), decoded straight back through `OggTrack` to prove the frame count — all four came back exact (88 200 / 88 200 and 176 400 / 176 400 frames), 27–61 KB each.
- `ProceduralAudio` loads them at startup and puts each through its bake's own law and voices: the shot and the burst through the blast's compressor and a 0.95 peak (not `Loudness` — its `tanh` is what you called "digital" on #389), the release peak-normalised to 0.9; then the same `VoiceRing`s, distance law and pitch jitter as before. **The landing keeps its colour ladder** by resampling the one file a whole tone per colour (the middle colour plays it as rendered, the ends an octave either way) times the material's transposition; what a file cannot carry is #314's per-material ring, partials and sub — every material now lands with your chosen timbre, shifted.
- **A missing file leaves the bake standing, silently** — delete any one of the four and that sound is procedural again.
- Smoke-tested: the Game launches with them, plays a level, the save is byte-identical.

**Not done, deliberately:** the victory fanfare. `ProceduralMusic.PlayVictory` is scaled by the score and answers `TryGetFanfare` for the star chime's key and tempo (#158), which a fixed recording cannot; `victory-fanfare-a-11` waits in the picks for a design that keeps those. The six alternates per chosen prompt are in `C:\Users\panrd\AI\sfx\out\482-final\index.html` if any of the four wears thin.

**Next is your ear in play:** shoot, land, clear a group, finish a level. If a sound sits wrong against the others, say which — the level is one constant per sound now, and a re-prompt is six seconds a render.

**Comment by AntoninPrazsky** [comment]
**The victory fanfare is a recording now — merge 3ad2d99 — on the owner's ruling: "the recording is the bed and the chime tunes to it."**

- `Game/Sfx/victory-fanfare.ogg` is `victory-fanfare-a-11`, stereo, with `ROOT=60` (C4) and `BPM=123.0` Vorbis comments. `MusicBake --sfx <wav> victory-fanfare --music` measured them: the key by a Krumhansl-profile match on the piece's chroma (C major 0.73, G 0.68, E 0.28 — C and G share every note but one, so the chime's C triad is consonant either way), the tempo by an autocorrelation of the onset flux; `--root`/`--bpm` override an estimate the ear disagrees with.
- `GameMusic` reads the tags into a `FanfareShape` and hands the decoded recording to `ProceduralMusic`, which plays it through the bake's own path — a completed task into the same realisation, clock, fades and fireworks ducking — so `TryGetFanfare` answers the star chime's key and beat exactly as before (#158): the stars land one beat apart on the recording's grid, retuned to its root.
- The defeat fanfare stays procedural; a victory file without its tags is not played and the bake stands.
- Smoke-tested on the result page with the celebration (`result celebrate`): no `[music]` fault, the save byte-identical.

**What a recording gives up:** the bake grew with the score (length and density); the recording is the same piece every time — which is what a win jingle should be, per the #229 ruling.

**Next is the owner's ear:** clear a level and listen to the fanfare with the stars landing on it. If the chime sounds out of tune, the key estimate was wrong: `--root <midi>` rewrites the tag in one run.

**Comment by AntoninPrazsky** [comment]
**Owner's ear in play (2026-09-21): all five sound good; the fanfare was "really very short".** Twelve longer renders — the original prompt and a two-phrase one, at 8 s and 12 s, three seeds each, all filling their length with music and ringing out, none clipping — and the verdict: **`fanfare-a-8s-11`** (the original prompt at 8 s); the 12 s ones "sound strange", the two-phrase ones "as if notes skipped… a little funereal". Lesson for the prompts: stretch the duration, leave the prompt alone — a more elaborate musical brief comes back unnatural.

**On `main` — merge 8158ff2:** `Game/Sfx/victory-fanfare.ogg` is that render, ROOT 60 (C major 0.70; F 0.58, A 0.51 behind it), BPM 161.5. The tempo estimate is coarse on a few seconds of brass — the 4 s render of the same piece measured 123 — so the stars pace at 161.5 until an ear says they rush, in which case `--bpm 120` rewrites the tag. Smoke-tested on the result page, save byte-identical.

Left for the owner: the fanfare with the stars landing on it, in play. Then this issue can close.

**Comment by AntoninPrazsky** [comment]
**Owner (2026-09-21): the 8 s brass fanfare "doesn't sound very modern and isn't 12 seconds."** Three modern directions at 12 s, three seeds each — synth-brass EDM, a cinematic hybrid (synth horns over electronic drums, a motif building three bars into a drop), pop-electronic arps — all filling 10–12 s, none clipping. **Pick: `fanfare-m2-12s-12`** (the hybrid).

**On `main` — merge a76d218:** `Game/Sfx/victory-fanfare.ogg` is that render, 12.0 s, ROOT 69 (A major at 0.90 — the surest key of any render measured; D 0.71, E 0.29), BPM 117.5, so the stars land 0.51 s apart on its grid. Smoke-tested on the result page, save byte-identical. Lesson recorded: name the style explicitly ("contemporary, electronic drums, synth"; negative "orchestral, classical, funeral, slow") — the brass prompts sounded dated to the owner however they were stretched.

Left for the owner: the fanfare with the stars in play; then close.

**Comment by AntoninPrazsky** [comment]
**Closed on the owner's verdict in play (2026-09-21): "Zní to dobře."** Five generated sounds ship from `Game/Sfx` — the shot, a landing, a release, the firework's burst and the modern twelve-second victory fanfare the star chime tunes to — with the procedural bakes standing behind any missing file. Pipeline: `C:\Users\panrd\AI\sfx` (renders, prompts, sidecars) and `Tools/MusicBake --sfx`; the record is under "The sound" in `docs/game-feedback.md` and in the `local-ai` skill. Still open beside it: #442 (the reference-only research this superseded) and the licence's registration and attribution clauses before a release, which are the owner's reading.

**Comment by AntoninPrazsky** [comment]
One consequence already heard in play (2026-09-21): the fireworks' launch whoosh now reads too loud against the bursts — filed as #498. The burst's recording is compressed and peak-normalised at load where the procedural one was driven to 0.30 RMS, so the balance between the procedural launch and the generated burst moved with this change and was never measured side by side.
